### PR TITLE
[RISCV] Add FPR16 regbank and start legalizing f16 operations for Zfh.

### PR DESCRIPTION
**Pull request by topperc**

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -371,17 +371,25 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
...
 
-  getActionDefinitionsBuilder({G_FADD, G_FSUB, G_FMUL, G_FDIV, G_FMA, G_FNEG,
-                               G_FABS, G_FSQRT, G_FMAXNUM, G_FMINNUM})
-      .legalIf(typeIsScalarFPArith(0, ST));
+  auto &FPArithActions = getActionDefinitionsBuilder(
+                             {G_FADD, G_FSUB, G_FMUL, G_FDIV, G_FMA, G_FNEG,
+                              G_FABS, G_FSQRT, G_FMAXNUM, G_FMINNUM})
+                             .legalIf(typeIsScalarFPArith(0, ST));
+  // TODO: Fold this into typeIsScalarFPArith.
+  if (ST.hasStdExtZfh())
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:379]: Doesn't it work?
```
static LegalityPredicate typeIsScalarFPArith(unsigned TypeIdx,
                                             const RISCVSubtarget &ST) {
  return [=, &ST](const LegalityQuery &Query) {
    return Query.Types[TypeIdx].isScalar() &&
           ((ST.hasStdExtF() && Query.Types[TypeIdx].getSizeInBits() == 32) ||
            (ST.hasStdExtD() && Query.Types[TypeIdx].getSizeInBits() == 64) ||
            (ST.hasStdExtZfh() && Query.Types[TypeIdx].getSizeInBits() == 16));
  };
}
```

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:fpr16 = COPY $f10_h
+    ; CHECK-NEXT: [[FSQRT_H:%[0-9]+]]:fpr16 = nofpexcept FSQRT_H [[COPY]], 7
+    ; CHECK-NEXT: $f10_h = COPY [[FSQRT_H]]
+    ; CHECK-NEXT: PseudoRET implicit $f10_h
+    %0:fprb(s16) = COPY $f10_h
+    %1:fprb(s16) = G_FSQRT %0
+    $f10_h = COPY %1(s16)
+    PseudoRET implicit $f10_h
+
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir:193]: Missing tests for G_FMAXNUM/G_FMINNUM/G_COPYSIGN.

**Review by dtcxzyw — COMMENTED** [review]
Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.

Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:379]: It should, but it requires many more tests to cover every other call to typeIsScalarFPArith

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir (hunk 2 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir:193]: Oops. I naively assumed that all the f32 tests were together when I copied this over.

**Comment by topperc** [comment]
> Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.

Why is that blocking?

> Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?

Yes.

**Comment by dtcxzyw** [comment]
> > Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.
> 
> Why is that blocking?

As we widen the type from f16 to f32, we need to implement G_FPTRUNC/G_FPEXT/G_ANYEXT for s16<->s32.

> 
> > Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?
> 
> Yes.

Great! TBH I don't have enough bandwidth to work on this. I will post my existing code later. Hope this helps you.

**Comment by dtcxzyw** [comment]
> > > Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.
> > 
> > 
> > Why is that blocking?
> 
> As we widen the type from f16 to f32, we need to implement G_FPTRUNC/G_FPEXT/G_ANYEXT for s16<->s32.
> 
> > > Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?
> > 
> > 
> > Yes.
> 
> Great! TBH I don't have enough bandwidth to work on this. I will post my existing code later. Hope this helps you.

See https://github.com/dtcxzyw/llvm-project/commit/7948041a4eaf9f89247c3f75b8afb2a5b998f642

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Wait for an additional approval from other reviewers.

**Review by arsenm — APPROVED** [review]

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Pull request merged**